pybind11/how-to.rst

```diff
@@ -645,110 +781,151 @@ Following :ref:`this rule <style-guide-pybind11-declare-template-wrappers>` we d
+As its name suggests, the former is needed when you want to inherit from a base class defined in another module; this must be called before the ``py::class_`` instantiation of the derived class.
+``addSignatureDependency`` declares that the external type is used only in function or method signatures.
+Regardless of when it is called, these dependencies will be imported after all local types are declared and before any definition callbacks are run.
 
-Adding these import statements in the C++ wrapper, rather than relying on import order in ``__init__`` in Python, follows :ref:`our rule on import <style-guide-pybind11-import>`.
+Circular *inheritance* dependencies are impossible in both C++ and Python, but circular signature dependencies are relatively common *within a single library* in C++ (via forward declarations) and quite possible in Python (via function-level imports and duck-typing).
+In pybind11 wrappers, circular signature dependencies are more of a problem than they are in either language independently.
+Using ``WrapperCollection`` solves that problem within a single module (in which all wrappers are added to the same ``WrapperCollection`` instance).
+It also does its best to work around circular dependencies between different modules, but this relies on the details of how Python handles circular imports, making it very hard to guarantee correct behavior in all cases.
+Circular dependencies between modules should be avoided whenever possible.
 
-Finished wrappers (C++ part)
-^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Binary Submodules
+^^^^^^^^^^^^^^^^^
 
-The end results for the C++ part of the wrappers (see next for the Python part) are ``exampleTwo.cc``:
+If the desired namespace for some symbols involves a subpackage nested below the level at which the ``WrapperCollection`` was defined, it's usually best to just define an entirely independent module for that subpackage.
 
-.. code-block:: cpp
+It's also possible to create a submodule within the same compilied module, however, and this can be necessary when the classes in the subpackage have circular dependencies with those in the main package or other subpackages.
 
-    #include "pybind11/pybind11.h"
+.. note::
 
-    #include "lsst/TMPL/ExampleTwo.h"
+    Binary submodules are complex and make the organization of a package's code difficult to understand.
+    Completely independent regular submodules should be used unless binary submodules are necessary to deal with circular dependencies.
+
+The module names and file/directory structure can be quite confusing in this case, so we'll look at a very concrete example.
+Let's imagine we have a package ``lsst.foo``, with a normal package-level
+module ``_foo``.
+That implies we have the following files:
+::
+    lsst/
+        foo/
+            __init__.py
+            _foo.cc
+            SConscript
```

Comment:
@jonathansick, sphinx is warning about "unexpected indentation" in this code block, possibly because it thinks this is Python code.  What should I be doing instead?

(note that it does actually render this correctly, and it did not when I was trying to use `.. code-block::` for it).